File: src/main/java/org/example/docvideoplay/service/DeckService.java
package org.example.docvideoplay.service;

import org.example.docvideoplay.entity.Deck;
import org.example.docvideoplay.entity.Card;

import java.util.List;

/**
 * Service for managing decks and their associated cards.
 * Handles deck creation, retrieval, and card management.
 */
public interface DeckService {
    
    /**
     * Create a new deck for a user.
     * 
     * @param userId The ID of the user creating the deck
     * @param name The name of the deck
     * @param description The description of the deck
     * @param isPublic Whether the deck is public
     * @return The created Deck entity
     */
    Deck createDeck(Long userId, String name, String description, Boolean isPublic);
    
    /**
     * Create a new deck with custom review options.
     * 
     * @param userId The ID of the user creating the deck
     * @param name The name of the deck
     * @param description The description of the deck
     * @param isPublic Whether the deck is public
     * @param newCardsPerDay Number of new cards per day
     * @param maxReviewsPerDay Maximum reviews per day
     * @param easyInterval Interval for easy cards
     * @param easyBonus Bonus multiplier for easy answers
     * @param intervalModifier Modifier for all intervals
     * @param startingEase Starting ease factor
     * @param steps Number of learning steps
     * @return The created Deck entity
     */
    Deck createDeckWithOptions(Long userId, String name, String description, Boolean isPublic,
                              Integer newCardsPerDay, Integer maxReviewsPerDay, Integer easyInterval,
                              Double easyBonus, Double intervalModifier, Double startingEase, Integer steps);
    
    /**
     * Retrieve a specific deck by ID.
     * 
     * @param deckId The ID of the deck
     * @return The deck entity
     * @throws IllegalArgumentException if deck not found
     */
    Deck getDeckById(Long deckId);
    
    /**
     * Retrieve a specific deck by ID for a specific user.
     * 
     * @param deckId The ID of the deck
     * @param userId The ID of the user who owns the deck
     * @return The deck entity
     * @throws IllegalArgumentException if deck not found or not owned by user
     */
    Deck getDeckById(Long deckId, Long userId);
    
    /**
     * Update an existing deck.
     * 
     * @param deckId The ID of the deck
     * @param name The new name
     * @param description The new description
     * @param isPublic The new public status
     * @param userId The ID of the user who owns the deck
     * @return The updated Deck entity
     * @throws IllegalArgumentException if deck not found or not owned by user
     */
    Deck updateDeck(Long deckId, String name, String description, Boolean isPublic, Long userId);
    
    /**
     * Update an existing deck with custom review options.
     * 
     * @param deckId The ID of the deck
     * @param name The new name
     * @param description The new description
     * @param isPublic The new public status
     * @param newCardsPerDay Number of new cards per day
     * @param maxReviewsPerDay Maximum reviews per day
     * @param easyInterval Interval for easy cards
     * @param easyBonus Bonus multiplier for easy answers
     * @param intervalModifier Modifier for all intervals
     * @param startingEase Starting ease factor
     * @param steps Number of learning steps
     * @param userId The ID of the user who owns the deck
     * @return The updated Deck entity
     * @throws IllegalArgumentException if deck not found or not owned by user
     */
    Deck updateDeckWithOptions(Long deckId, String name, String description, Boolean isPublic,
                              Integer newCardsPerDay, Integer maxReviewsPerDay, Integer easyInterval,
                              Double easyBonus, Double intervalModifier, Double startingEase, Integer steps,
                              Long userId);
    
    /**
     * Delete a deck and all its associated cards.
     * 
     * @param deckId The ID of the deck to delete
     * @param userId The ID of the user who owns the deck
     * @throws IllegalArgumentException if deck not found or not owned by user
     */
    void deleteDeck(Long deckId, Long userId);
    
    /**
     * Retrieve all decks for a specific user.
     * 
     * @param userId The ID of the user who created the decks
     * @return List of decks ordered by creation date
     */
    List<Deck> getDecksByUser(Long userId);
    
    /**
     * Retrieve all public decks.
     * 
     * @return List of public decks ordered by creation date
     */
    List<Deck> getPublicDecks();
    
    /**
     * Retrieve all public decks created by a specific user.
     * 
     * @param userId The ID of the user who created the decks
     * @return List of public decks ordered by creation date
     */
    List<Deck> getPublicDecksByUser(Long userId);
    
    /**
     * Add a card to a deck.
     * 
     * @param deckId The ID of the deck
     * @param card The card to add
     * @param userId The ID of the user who owns the deck
     * @return The updated Deck entity
     * @throws IllegalArgumentException if deck not found or not owned by user
     */
    Deck addCardToDeck(Long deckId, Card card, Long userId);
    
    /**
     * Remove a card from a deck.
     * 
     * @param deckId The ID of the deck
     * @param cardId The ID of the card to remove
     * @param userId The ID of the user who owns the deck
     * @return The updated Deck entity
     * @throws IllegalArgumentException if deck not found, card not found, or not owned by user
     */
    Deck removeCardFromDeck(Long deckId, Long cardId, Long userId);
    
    /**
     * Retrieve all cards in a deck.
     * 
     * @param deckId The ID of the deck
     * @param userId The ID of the user who owns the deck
     * @return List of cards in the deck
     * @throws IllegalArgumentException if deck not found or not owned by user
     */
    List<Card> getCardsInDeck(Long deckId, Long userId);
    
    /**
     * Retrieve all cards in a public deck.
     * 
     * @param deckId The ID of the public deck
     * @return List of cards in the deck
     * @throws IllegalArgumentException if deck not found or not public
     */
    List<Card> getCardsInPublicDeck(Long deckId);
    
    /**
     * Search decks by name or description.
     * 
     * @param searchText The text to search for
     * @param userId The ID of the user performing the search
     * @return List of decks matching the search criteria
     */
    List<Deck> searchDecks(String searchText, Long userId);
    
    /**
     * Get the count of cards in a deck.
     * 
     * @param deckId The ID of the deck
     * @param userId The ID of the user who owns the deck
     * @return The number of cards in the deck
     * @throws IllegalArgumentException if deck not found or not owned by user
     */
    long getCardCountInDeck(Long deckId, Long userId);
    
    /**
     * Duplicate a deck and its cards.
     * 
     * @param deckId The ID of the deck to duplicate
     * @param userId The ID of the user who will own the duplicate deck
     * @param newName The name of the duplicate deck
     * @return The duplicated Deck entity
     * @throws IllegalArgumentException if deck not found
     */
    Deck duplicateDeck(Long deckId, Long userId, String newName);
}

File: frontend/src/services/deckService.js
import { useApiService } from '../composables/useApiService'

/**
 * Service for deck management operations
 * Handles communication with the backend DeckApi
 */
export function useDeckService() {
  const { apiService } = useApiService()

  /**
   * Create a new deck
   * @param {Object} deckData - Deck creation data
   * @param {string} deckData.name - Deck name
   * @param {string} deckData.description - Optional deck description
   * @param {boolean} deckData.isPublic - Optional public visibility
   * @returns {Promise<Object>} Created deck data
   */
  const createDeck = async (deckData) => {
    try {
      const response = await apiService.post('/decks', deckData)
      return response.data
    } catch (error) {
      console.error('Failed to create deck:', error)
      throw new Error(error.response?.data?.message || 'Failed to create deck')
    }
  }

  /**
   * Get all decks for the current user
   * @returns {Promise<Array>} List of user's decks
   */
  const getAllDecks = async () => {
    try {
      const response = await apiService.get('/decks')
      return response.data || []
    } catch (error) {
      console.error('Failed to fetch decks:', error)
      throw new Error(error.response?.data?.message || 'Failed to fetch decks')
    }
  }

  /**
   * Get a specific deck by ID
   * @param {number} id - Deck ID
   * @returns {Promise<Object>} Deck data
   */
  const getDeck = async (id) => {
    try {
      const response = await apiService.get(`/decks/${id}`)
      return response.data
    } catch (error) {
      console.error(`Failed to fetch deck ${id}:`, error)
      if (error.response?.status === 404) {
        throw new Error('Deck not found')
      }
      throw new Error(error.response?.data?.message || 'Failed to fetch deck')
    }
  }

  /**
   * Update a deck
   * @param {number} id - Deck ID
   * @param {Object} updateData - Update data
   * @param {string} updateData.name - Updated deck name
   * @param {string} updateData.description - Updated deck description
   * @param {boolean} updateData.isPublic - Updated public visibility
   * @returns {Promise<Object>} Updated deck data
   */
  const updateDeck = async (id, updateData) => {
    try {
      const response = await apiService.put(`/decks/${id}`, updateData)
      return response.data
    } catch (error) {
      console.error(`Failed to update deck ${id}:`, error)
      if (error.response?.status === 404) {
        throw new Error('Deck not found')
      }
      throw new Error(error.response?.data?.message || 'Failed to update deck')
    }
  }

  /**
   * Delete a deck
   * @param {number} id - Deck ID
   * @returns {Promise<void>}
   */
  const deleteDeck = async (id) => {
    try {
      await apiService.delete(`/decks/${id}`)
    } catch (error) {
      console.error(`Failed to delete deck ${id}:`, error)
      if (error.response?.status === 404) {
        throw new Error('Deck not found')
      }
      throw new Error(error.response?.data?.message || 'Failed to delete deck')
    }
  }

  /**
   * Get all cards in a deck
   * @param {number} id - Deck ID
   * @returns {Promise<Array>} List of cards in the deck
   */
  const getCardsInDeck = async (id) => {
    try {
      const response = await apiService.get(`/decks/${id}/cards`)
      return response.data || []
    } catch (error) {
      console.error(`Failed to fetch cards in deck ${id}:`, error)
      if (error.response?.status === 404) {
        throw new Error('Deck not found')
      }
      throw new Error(error.response?.data?.message || 'Failed to fetch deck cards')
    }
  }

  /**
   * Add a card to a deck
   * @param {number} deckId - Deck ID
   * @param {number} cardId - Card ID to add
   * @returns {Promise<Object>} Updated deck data
   */
  const addCardToDeck = async (deckId, cardId) => {
    try {
      const response = await apiService.post(`/decks/${deckId}/cards`, { cardId })
      return response.data
    } catch (error) {
      console.error(`Failed to add card to deck ${deckId}:`, error)
      if (error.response?.status === 404) {
        throw new Error('Deck not found')
      }
      throw new Error(error.response?.data?.message || 'Failed to add card to deck')
    }
  }

  /**
   * Remove a card from a deck
   * @param {number} deckId - Deck ID
   * @param {number} cardId - Card (highlight) ID to remove
   * @returns {Promise<Object>} Updated deck data
   */
  const removeCardFromDeck = async (deckId, cardId) => {
    try {
      const response = await apiService.delete(`/decks/${deckId}/cards/${cardId}`)
      return response.data
    } catch (error) {
      console.error(`Failed to remove card from deck ${deckId}:`, error)
      if (error.response?.status === 404) {
        throw new Error('Deck or card not found')
      }
      throw new Error(error.response?.data?.message || 'Failed to remove card from deck')
    }
  }

  /**
   * Get all public decks
   * @returns {Promise<Array>} List of public decks
   */
  const getPublicDecks = async () => {
    try {
      const response = await apiService.get('/decks/public')
      return response.data || []
    } catch (error) {
      console.error('Failed to fetch public decks:', error)
      throw new Error(error.response?.data?.message || 'Failed to fetch public decks')
    }
  }

  /**
   * Duplicate a deck
   * @param {number} deckId - Deck ID to duplicate
   * @param {string} newName - Name for the duplicated deck
   * @returns {Promise<Object>} Duplicated deck data
   */
  const duplicateDeck = async (deckId, newName) => {
    try {
      const response = await apiService.post(`/decks/${deckId}/duplicate`, { newName })
      return response.data
    } catch (error) {
      console.error(`Failed to duplicate deck ${deckId}:`, error)
      if (error.response?.status === 404) {
        throw new Error('Deck not found')
      }
      throw new Error(error.response?.data?.message || 'Failed to duplicate deck')
    }
  }

  /**
   * Search decks by name or description
   * @param {string} query - Search query
   * @returns {Promise<Array>} List of matching decks
   */
  const searchDecks = async (query) => {
    try {
      const response = await apiService.get('/decks/search', {
        params: { q: query }
      })
      return response.data || []
    } catch (error) {
      console.error('Failed to search decks:', error)
      throw new Error(error.response?.data?.message || 'Failed to search decks')
    }
  }

  return {
    createDeck,
    getAllDecks,
    getDeck,
    updateDeck,
    deleteDeck,
    getCardsInDeck,
    addCardToDeck,
    removeCardFromDeck,
    getPublicDecks,
    duplicateDeck,
    searchDecks
  }
}
